### PR TITLE
Queue improvements

### PR DESCRIPTION
**Pull request by jettero**
* removed diskcache, which uses sqlite3 (I feel like that's asking for trouble)
* refactored the http_event_collector options processing (more DRY is good)
* added some options @fossam wanted (compression, location, size of the diskqueue)
* fixed a missing log.error() when send fails without being an urllib3 exception
* found a major bug that prevented flushing the diskqueue; which without flushing is just a catalog of failure that never clears

### CODE REVIEW

File: hubblestack/hec/dq.py [hunk 1]
```diff
@@ -1,129 +1,292 @@
...
+        return _bz2(dat)
 
-    # copy *some* of the operator overloading up from the q
-    def __iter__(self):
-        return self.q.__iter__()
+    def decompress(self, dat):
+        if dat.startswith('BZ'):
+            try:
+                return bz2.BZ2Decompressor().decompress(dat)
+            except IOError:
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: I think of the queue as transient and as a last resort. If things fail pulling items from the queue, I think it's best to ignore it. _Normally_ Hubble shouldn't be using the queue anyway.

If this is the wrong philosophy, I may have treated other things too loosely.

File: hubblestack/hec/dq.py [hunk 2]
```diff
@@ -1,129 +1,292 @@
...
-    def __contains__(self, thing):
-        return thing in self.q
+    def _mkdir(self, partial=None):
+        d = self.directory
+        if partial is not None:
+            d = os.path.join(d, partial)
+        if not os.path.isdir(d):
+            try: os.makedirs(d)
+            except OSError as e:
+                if e.errno == errno.EEXIST and os.path.isdir(d): pass
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: If we can't make the dir, and the dir isn't a dir and the file exists.... then the name is taken by a file that is not a directory. It's an unresolvable condition, or at least it's one that requires human intervention.

File: hubblestack/hec/dq.py [hunk 3]
```diff
@@ -1,129 +1,292 @@
...
+    def accept(self, item):
+        if len(item) + self.sz > self.size:
+            return False
+        return True
+
+    def put(self, item):
+        self.check_type(item)
+        if not self.accept(item):
+            raise QueueCapacityError('refusing to accept item due to size')
+        fanout,remainder = self._fanout('{0}.{1}'.format(int(time.time()), self.cn))
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: we need to get the linting working. for the basic whitespace formatting at the very least.

File: hubblestack/hec/dq.py [hunk 4]
```diff
@@ -1,129 +1,292 @@
...
+        except QueueCapacityError:
+            self.dq.put(item)
+
+    def peek(self):
+        r = self.mq.peek()
+        if r is None:
+            r = self.dq.peek()
+        return r
+
+    def unget(self, msg):
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: lol. I attempted to document everything. though, we really only use the DiskQueue, which doesn't have an unget method (it's not clear how that should work, so I never finished the thought).

File: hubblestack/hec/dq.py [hunk 5]
```diff
@@ -1,129 +1,292 @@
...
+            try:
+                return [ int(i) for i in x.split('.') ]
+            except:
+                pass
+            return x
+        for path, dirs, files in sorted(os.walk(self.directory)):
+            for fname in [ os.path.join(path, f) for f in sorted(files, key=_k) ]:
+                yield fname
+
+    def _count(self):
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: originally I wrote .sz and .cn as boring @properties that each traversed the diskqueue. As a first-step optimization, I decided to pre-compute it on each insert...

 I pictured the queue staying quite small normally and decided not to continue optimizing it. It probably doesn't matter very much unless the hec isn't accepting messages anyway and that's going to be slow enough and rare enough that the recounts will only be once every second or so...

That's my thinking anyway. I could obviously make this quite a lot better.

File: hubblestack/hec/dq.py [hunk 6]
```diff
@@ -1,129 +1,292 @@
...
+    @property
+    def msz(self):
+        return self.sz + max(0, self.cn-1)
+
+    def __len__(self):
+        return self.msz
+
+class DiskBackedQueue:
+    def __init__(self, directory, mem_size=DEFAULT_MEMORY_SIZE,
+        disk_size=DEFAULT_DISK_SIZE, ok_types=OK_TYPES, fresh=False):
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/hec/dq.py]: oh, I really don't like it when the args line up with the def line. it makes it look like it's not a list continuation.  I mean, I'll do it, but I'd really rather not.  yeah?

File: hubblestack/hec/dq.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (basepi) [reply hubblestack/hec/dq.py]: I think I'm ok with it, on further thought.

File: hubblestack/hec/dq.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (basepi) [reply hubblestack/hec/dq.py]: The problem is that python is an indent-driven language. So using a 4-indent is super confusing. The other solution is adding a docstring, as a visible separator. I don't care either way, as long as it's not a 4-space indent followed by a different block at the same 4-space indent. Does that make sense?

**Comment by basepi** [comment]
Thanks for the additional docstrings. Just one unresolved comment above from my end.

File: hubblestack/hec/dq.py [hunk 7]
```diff
@@ -1,129 +1,376 @@
...
+    def msz(self):
+        ''' The size of the queue as it would be returned from getz() iff getz had no size limit '''
+        return self.sz + max(0, len(self.sep) * (self.cn -1))
+
+    def __len__(self):
+        return self.msz
+
+class DiskBackedQueue:
+    def __init__(self, directory, mem_size=DEFAULT_MEMORY_SIZE,
+                            disk_size=DEFAULT_DISK_SIZE, ok_types=OK_TYPES, fresh=False):
```
Comment (basepi) [inline hubblestack/hec/dq.py]: I think the new indentation must have been an accident, or else you're trolling me.

```
class DiskBackedQueue:
        def __init__(self, directory, mem_size=DEFAULT_MEMORY_SIZE,
                disk_size=DEFAULT_DISK_SIZE, ok_types=OK_TYPES, fresh=False):
```
is what we're looking for.

**Pull request merged**